docs/design/2022-04-15-multi-schema-change.md

```diff
@@ -0,0 +1,126 @@
+        SubJobs: []*SubJob {
+            &SubJob {
+                Type: ActionAddColumn,
+                Args: ...
+            },
+            &SubJob {
+                Type: ActionModifyColumn,
+                Args: ...
+            },
+        }
+    } 
+}
+```
+
+In this way, we pack multiple schema changes into one job. Just like any other jobs, it enters the DDL job queue stored in TiKV and waits for a suitable worker to pick it up.
+
+In normal cases, the worker executes the sub-jobs one by one serially as if they are plain jobs. However, the thing gets complex in abnormal cases.
+
+To ensure the atomicity of a Multi-Schema Change execution, we need to maintain the changing schema object states carefully. Take the above SQL as an example, if the second sub-job `MODIFY COLUMN a CHAR(255)` failed for some reason, the first sub-job should be able to rollback its changes(rollback the added column `b`).
+
+This requirement means we cannot simply public the schema object when a sub-job is finished. Instead, it should stay in a state which is still unnoticeable to users, and wait for the other sub-jobs, finally public all at once when all the sub-jobs are confirmed to be successful. This method is kind of like 2PC: the "commit" can only be started when "prewrites" are complete.
+
+Here is the table of states that should stay for different DDLs. Note the "Next State" means the changes are noticeable to the users.
+
+| DDL (Schema Change)     | Unnoticeable State | Next State |
+|-------------------------|--------------------|------------|
+| Add Column              | Write-Reorg        | Public     |
+| Add Index               | Write-Reorg        | Public     |
+| Drop Column             | Public             | Write-Only |
+| Drop Index              | Public             | Write-Only |
+| Non-reorg Modify Column | Public             | Public     |
+| Reorg Modify Column     | Write-Reorg        | Public     |
+
+To achieve this behavior, we introduce a flag in the sub-job named "non-revertible". This flag is set when a schema object has stepped to the last unnoticeable state. When all sub-jobs become non-revertible, all the related schema objects step to the next state in one transaction. After that, the sub-jobs are executed serially to do the rest.
+
+On the other hand, if there is an error returned by any sub-job before all of them become non-revertible, the whole job is switched to a `rollingback` state. For the executed sub-jobs, we set them to `cancelling`; for the sub-job that have not been executed, we set them to `cancelled`.
+
+Finally, we consider the extreme case: an error occurs while all the sub-jobs are non-revertible. In this situation, we tend to believe that the error can be solved in a trivial way like retrying. This behavior is compatible with the current DDL implementation. Take `DROP COLUMN` as an example, once the column steps to the "Write-Only" state, there is no way to cancel this job.
+
+## Compatibility
```

Comment:
Please briefly explain the compatibility problems encountered during the rolling upgrade
The owner is a new TiDB and the old TiDB receives the `multi-schema-change` related DDL.

Reply:
Updated.